File: website/vulnerable-demo.py
"""
INTENTIONALLY VULNERABLE CODE FOR CODEQL TESTING
DO NOT USE IN PRODUCTION

This file contains security vulnerabilities to demonstrate
GitHub CodeQL code scanning capabilities.
"""

import os
import pickle
import subprocess
import sqlite3
from flask import Flask, request, render_template_string, redirect

app = Flask(__name__)

# ============================================================
# VULNERABILITY 1: SQL Injection
# CodeQL should detect: py/sql-injection
# ============================================================
@app.route('/user')
def get_user():
    user_id = request.args.get('id')
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    # VULNERABLE: User input directly concatenated into SQL query
    query = "SELECT * FROM users WHERE id = '" + user_id + "'"
    cursor.execute(query)
    
    return str(cursor.fetchall())


# ============================================================
# VULNERABILITY 2: Command Injection
# CodeQL should detect: py/command-line-injection
# ============================================================
@app.route('/ping')
def ping_host():
    host = request.args.get('host')
    
    # VULNERABLE: User input passed directly to shell command
    result = os.popen('ping -c 4 ' + host).read()
    return result


# ============================================================
# VULNERABILITY 3: Command Injection (subprocess)
# CodeQL should detect: py/command-line-injection
# ============================================================
@app.route('/execute')
def execute_command():
    cmd = request.args.get('cmd')
    
    # VULNERABLE: User input in shell command
    result = subprocess.run(cmd, shell=True, capture_output=True)
    return result.stdout.decode()


# ============================================================
# VULNERABILITY 4: Server-Side Template Injection (SSTI)
# CodeQL should detect: py/template-injection
# ============================================================
@app.route('/greet')
def greet():
    name = request.args.get('name', 'World')
    
    # VULNERABLE: User input in template without sanitization
    template = f"<h1>Hello, {name}!</h1>"
    return render_template_string(template)


# ============================================================
# VULNERABILITY 5: Cross-Site Scripting (XSS)
# CodeQL should detect: py/reflective-xss
# ============================================================
@app.route('/search')
def search():
    query = request.args.get('q', '')
    
    # VULNERABLE: User input reflected in response
    return f"""
    <html>
        <body>
            <h1>Search Results</h1>
            <p>You searched for: {query}</p>
        </body>
    </html>
    """


# ============================================================
# VULNERABILITY 6: Path Traversal
# CodeQL should detect: py/path-injection
# ============================================================
@app.route('/file')
def read_file():
    filename = request.args.get('name')
    
    # VULNERABLE: User input used directly in file path
    filepath = os.path.join('./uploads/', filename)
    with open(filepath, 'r') as f:
        return f.read()


# ============================================================
# VULNERABILITY 7: Insecure Deserialization
# CodeQL should detect: py/unsafe-deserialization
# ============================================================
@app.route('/load', methods=['POST'])
def load_data():
    data = request.data
    
    # VULNERABLE: Deserializing untrusted data with pickle
    obj = pickle.loads(data)
    return str(obj)


# ============================================================
# VULNERABILITY 8: Hardcoded Credentials
# CodeQL should detect: py/hardcoded-credentials
# ============================================================
DATABASE_CONFIG = {
    'host': 'database.example.com',
    'username': 'admin',
    'password': 'SuperSecret123!',  # VULNERABLE: Hardcoded password
    'database': 'production'
}

API_KEY = "sk-1234567890abcdef"  # VULNERABLE: Hardcoded API key


# ============================================================
# VULNERABILITY 9: Open Redirect
# CodeQL should detect: py/url-redirection
# ============================================================
@app.route('/redirect')
def unsafe_redirect():
    url = request.args.get('url')
    
    # VULNERABLE: Redirecting to user-controlled URL
    return redirect(url)


# ============================================================
# VULNERABILITY 10: Weak Cryptography
# CodeQL should detect: py/weak-cryptographic-algorithm
# ============================================================
import hashlib

@app.route('/hash')
def hash_password():
    password = request.args.get('password')
    
    # VULNERABLE: Using MD5 for password hashing
    hashed = hashlib.md5(password.encode()).hexdigest()
    return {'hash': hashed}


# ============================================================
# VULNERABILITY 11: XML External Entity (XXE)
# CodeQL should detect: py/xxe
# ============================================================
from xml.etree.ElementTree import parse

@app.route('/parse-xml', methods=['POST'])
def parse_xml():
    # VULNERABLE: Parsing XML without disabling external entities
    tree = parse(request.stream)
    return str(tree.getroot().text)


# ============================================================
# VULNERABILITY 12: LDAP Injection
# CodeQL should detect: py/ldap-injection
# ============================================================
import ldap

@app.route('/ldap-search')
def ldap_search():
    username = request.args.get('username')
    
    # VULNERABLE: User input in LDAP filter
    conn = ldap.initialize('ldap://localhost')
    filter_str = f"(uid={username})"
    results = conn.search_s('dc=example,dc=com', ldap.SCOPE_SUBTREE, filter_str)
    return str(results)


# ============================================================
# Start Flask server
# ============================================================
if __name__ == '__main__':
    # VULNERABLE: Debug mode enabled in production
    app.run(debug=True, host='0.0.0.0')

File: website/vulnerable-demo.js
/**
 * INTENTIONALLY VULNERABLE CODE FOR CODEQL TESTING
 * DO NOT USE IN PRODUCTION
 * 
 * This file contains security vulnerabilities to demonstrate
 * GitHub CodeQL code scanning capabilities.
 */

const express = require('express');
const { exec } = require('child_process');
const fs = require('fs');
const mysql = require('mysql');

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// ============================================================
// VULNERABILITY 1: SQL Injection
// CodeQL should detect: js/sql-injection
// ============================================================
app.get('/user', (req, res) => {
    const userId = req.query.id;
    const connection = mysql.createConnection({
        host: 'localhost',
        user: 'root',
        password: 'password',
        database: 'users'
    });
    
    // VULNERABLE: User input directly concatenated into SQL query
    const query = "SELECT * FROM users WHERE id = '" + userId + "'";
    connection.query(query, (error, results) => {
        if (error) throw error;
        res.json(results);
    });
});

// ============================================================
// VULNERABILITY 2: Command Injection
// CodeQL should detect: js/command-line-injection
// ============================================================
app.get('/ping', (req, res) => {
    const host = req.query.host;
    
    // VULNERABLE: User input passed directly to shell command
    exec('ping -c 4 ' + host, (error, stdout, stderr) => {
        res.send(stdout || stderr);
    });
});

// ============================================================
// VULNERABILITY 3: Cross-Site Scripting (XSS)
// CodeQL should detect: js/xss, js/reflected-xss
// ============================================================
app.get('/search', (req, res) => {
    const searchTerm = req.query.q;
    
    // VULNERABLE: User input reflected in response without sanitization
    res.send(`<html><body>
        <h1>Search Results</h1>
        <p>You searched for: ${searchTerm}</p>
    </body></html>`);
});

// ============================================================
// VULNERABILITY 4: Path Traversal
// CodeQL should detect: js/path-injection
// ============================================================
app.get('/file', (req, res) => {
    const filename = req.query.name;
    
    // VULNERABLE: User input used directly in file path
    const filepath = './uploads/' + filename;
    fs.readFile(filepath, (err, data) => {
        if (err) {
            res.status(404).send('File not found');
        } else {
            res.send(data);
        }
    });
});

// ============================================================
// VULNERABILITY 5: Hardcoded Credentials
// CodeQL should detect: js/hardcoded-credentials
// ============================================================
const dbConfig = {
    host: 'database.example.com',
    username: 'admin',
    password: 'SuperSecret123!',  // VULNERABLE: Hardcoded password
    database: 'production'
};

// ============================================================
// VULNERABILITY 6: Insecure Randomness
// CodeQL should detect: js/insecure-randomness
// ============================================================
app.get('/token', (req, res) => {
    // VULNERABLE: Math.random() is not cryptographically secure
    const token = Math.random().toString(36).substring(2);
    res.json({ token: token });
});

// ============================================================
// VULNERABILITY 7: Missing Rate Limiting (Logic Issue)
// ============================================================
app.post('/login', (req, res) => {
    const { username, password } = req.body;
    // No rate limiting - allows brute force attacks
    authenticateUser(username, password)
        .then(user => res.json({ success: true, user }))
        .catch(err => res.status(401).json({ error: 'Invalid credentials' }));
});

// ============================================================
// VULNERABILITY 8: Prototype Pollution
// CodeQL should detect: js/prototype-polluting-assignment
// ============================================================
app.post('/config', (req, res) => {
    const userConfig = req.body;
    const config = {};
    
    // VULNERABLE: Merging user-controlled object without validation
    for (const key in userConfig) {
        config[key] = userConfig[key];
    }
    
    res.json({ message: 'Config updated', config });
});

// ============================================================
// VULNERABILITY 9: Regex DoS (ReDoS)
// CodeQL should detect: js/redos
// ============================================================
app.get('/validate-email', (req, res) => {
    const email = req.query.email;
    
    // VULNERABLE: Exponential backtracking regex
    const emailRegex = /^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$/;
    
    if (emailRegex.test(email)) {
        res.json({ valid: true });
    } else {
        res.json({ valid: false });
    }
});

// ============================================================
// VULNERABILITY 10: Server-Side Request Forgery (SSRF)
// CodeQL should detect: js/request-forgery
// ============================================================
const axios = require('axios');

app.get('/fetch', async (req, res) => {
    const url = req.query.url;
    
    // VULNERABLE: Fetching user-controlled URL without validation
    try {
        const response = await axios.get(url);
        res.send(response.data);
    } catch (error) {
        res.status(500).send('Error fetching URL');
    }
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});

function authenticateUser(username, password) {
    // Dummy function
    return Promise.reject(new Error('Not implemented'));
}

module.exports = app;
